api/src/wfl/module/covid.clj

```diff
@@ -23,6 +24,61 @@
 (defn create-covid-workload!
   [tx {:keys [source sink executor] :as request}])
 
+(defn start-covid-workload!
+  [{:keys [submission-id workspace] :as workload}])
+
+(defn ^:private snapshot-error-map
+  "Return default ExceptionInfo map for snapshot-related exceptions."
+  [{:keys [items sink] :as _workload} {:keys [id snapshot_id] :as _record}]
+  {:workspace (:workspace sink)
+   :table items
+   :id id
+   :snapshot_id snapshot_id})
+
+(defn ^:private try-get-snapshot-reference
+  "Attempt to fetch snapshot reference with id SNAPSHOT_REFERENCE_ID
+  from SINK-derived workspace."
+  [{:keys [sink] :as workload} {:keys [snapshot_reference_id] :as record}]
+  (let [workspace (:workspace sink)]
+    (try
+      (rawls/get-snapshot-reference workspace snapshot_reference_id)
+      (catch Throwable cause
+        (throw (ex-info "Rawls unable to fetch snapshot reference"
+                        (util/deep-merge
+                         (snapshot-error-map workload record)
+                         {:snapshot_reference_id snapshot_reference_id})
+                        cause))))))
+
+(defn ^:private snapshot-imported?
+  "Check if snapshot with id SNAPSHOT_ID in TDR has been imported to
+  SINK-derived workspace and can be successfully fetched."
+  [workload {:keys [snapshot_id snapshot_reference_id] :as record}]
+  {:pre [(some? snapshot_id)]}
+  (and (some? snapshot_reference_id)
+       (some? (try-get-snapshot-reference workload record))))
```

Comment:
Short-circuiting means that we'll try to get the snapshot reference from Rawls only if a `snapshot_reference_Id` is defined.

If it's defined in the DB, we'd hope that we can fetch the snapshot reference successfully… but there might be the possibility that a snapshot is successfully imported in the last run of the update loop and corrupted / deleted during the sleep.

If that happens and the fetch fails, should we…
- Throw the error in `try-get-snapshot-reference` as we do now, halting the processing?
- Log the error but do not throw so the snapshot import retries gracefully?

Reply:
If the snapshot has a snapshot reference, then isn't it already imported?

Reply:
Yes.  But this guards against the (probably low) chance of the snapshot reference being deleted in between import and this check, since they happen in different runs of the update loop.

This may be an overoptimization: I can revert to a simple check for the reference id if so.